### DISCUSSION TITLE
Gathering price info of each CSP (VM, NET, ... )

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#753] [Show and tell]
# Resource pricing info of each CSP 
- (1) VM resource 
- (2) Network resource

## (1) VM resource 
- on demand (price/hour)
- in US $

### AWS
- https://aws.amazon.com/ec2/pricing/on-demand/?nc1=h_ls
- AWS 서비스 가격 대량 쿼리 가능 (JSON 또는 CSV 파일 제공: AWS Price List Bulk API)
  - AWS SDK 또는 AWS CLI를 통해 AWS 서비스, 제품 및 요금에 대한 특정 정보를 쿼리 (AWS Price List Query API 사용)
  - https://docs.aws.amazon.com/ko_kr/awsaccountbilling/latest/aboutv2/price-changes.html 

### Alibaba
- https://www.alibabacloud.com/help/en/doc-detail/91486.html 
- API로 쿼리 가능 (https://business.ap-southeast-1.aliyuncs.com/?Action=GetPayAsYouGoPrice) (JSON응답)

### Azure
- API로 쿼리 가능 (https://prices.azure.com/api/retail/prices) (JSON응답)
- https://docs.microsoft.com/ko-kr/rest/api/cost-management/retail-prices/azure-retail-prices 

### GCP
- https://cloud.google.com/billing/v1/how-tos/catalog-api 
- API로 쿼리 가능 (https://cloudbilling.googleapis.com/v1/services) (JSON응답)

### Tencent
- https://intl.cloud.tencent.com/pricing/cvm?regionId=33&zoneId=330001&lang=en&pg=

### NCP
- API로 쿼리 가능 ({API_URL}/product/getProductPriceList) (JSON응답)
- https://api.ncloud-docs.com/docs/platform-listprice-getproductpricelist

### IBM
- https://cloud.ibm.com/vpc-ext/provision/vs

## (2) Network resource
This discussion is to share pricing of virtual network resources of each Cloud Service Provider.
This information is for reference and may not accurate and up to date.

Note that CB-Tumblebug uses basic network resources such as VPC network, subnet, and so on.

Please update this discussion or make a comment if you find any wrong information.

### AWS
  - https://aws.amazon.com/vpc/pricing/?nc1=h_ls
  - > There are no additional charges for creating and using an Amazon Virtual Private Cloud (VPC) itself. If you choose to create a Network Address Translation (NAT) gateway in your VPC, you are charged for each “NAT Gateway-hour" that your NAT gateway is provisioned and available. Amazon VPC Reachability Analyzer charges you each time you analyze connectivity between a given source and destination. If you choose to enable traffic mirroring, you will be charged hourly for each Elastic Network Interface (ENI) enabled with traffic mirroring. For specific pricing rates please read below.
  - **No charges for VPC itself**

### Alibaba
  - https://partners-intl.aliyun.com/help/doc-detail/127043.htm
  - > VPC is free of charge. However, you are charged for resources deployed in a VPC network.
  - **VPC is free of charg**

### Azure
  - https://azure.microsoft.com/pricing/details/virtual-network/
  - > Azure의 Virtual Network는 무료로 제공됩니다. 모든 구독은 모든 지역에서 최대 50개의 Virtual Network를 만들 수 있습니다. VNET 피어링은 두 가상 네트워크(같은 지역 또는 다른 지역에서)를 연결하며 개인 IP 주소(명목상의 요금이 부과됨)를 사용하여 두 네트워크 간에 트래픽을 라우팅하는 데 사용할 수 있습니다. 인바운드 및 아웃바운드 트래픽은 피어링된 네트워크의 양쪽 끝에서 요금이 청구됩니다. 가상 네트워크 내에서 실행되는 VPN Gateway 및 Application Gateway와 같은 네트워크 어플라이언스에도 요금이 부과됩니다.
  - **Azure의 Virtual Network는 무료로 제공**

### GCP
  - https://cloud.google.com/vpc/network-pricing
  - > No charge for ingress traffic. However there may be a charge for resource that processes ingress traffic. Services that process ingress traffic are as follows ...
  - Cannot find explicit notice. but **No charge for ingress traffic**

### Tencent
  - https://intl.cloud.tencent.com/product/vpc?lang=en&pg=#m_pricing
  - > Virtual Private Cloud (VPC) is free of charge. However, while using VPC, fees for other related products such as NAT Gateway, VPN Connection and Peering Connection may be incurred. 
  -  **Virtual Private Cloud (VPC) is free of charge**

### NCP
  - https://www.ncloud.com/product/networking/vpc
  - > VPC와 서브넷, NACL, ACG, Virtual Private Gateway는 무료입니다.
  - **VPC와 서브넷, NACL, ACG, Virtual Private Gateway는 무료**

### IBM
 - https://cloud.ibm.com/vpc-ext/provision/vpc
 - vpc 및 서브넷 추가에 의한 예상 비용 발생하지 않음. (참고: 클래식 버전인 Vlan은 과금 여부 확인하지 않음.)

**Comment by seokho-son** [comment]
@jihoon-seo 

IBM Cloud: Vlan 버전이 생성 과금이 있나봐요.

https://github.com/cloud-barista/cb-spider/issues/401#issuecomment-875297017

> VLan vs VPC 사용 비용
> VPC : VPC 당 사용료는 없지만, virtaulServer당 floating ip 1달러
> Vlan: 처음 생성되는 VLAN 사용시 사용료 없음. 그러나 사용자가 VLAN 생성시 25달러과금 인스턴스당 public IP비용 없음

**Comment by jihoon-seo** [comment]
From @seokho-son :

> - AWS 서비스 가격 대량 쿼리 가능 (JSON 또는 CSV 파일 제공: AWS Price List Bulk API)
> - AWS SDK 또는 AWS CLI를 통해 AWS 서비스, 제품 및 요금에 대한 특정 정보를 쿼리 (AWS Price List Query API 사용)
> - https://docs.aws.amazon.com/ko_kr/awsaccountbilling/latest/aboutv2/price-changes.html 

API
- https://docs.aws.amazon.com/ko_kr/awsaccountbilling/latest/aboutv2/using-pelong.html
- https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_pricing_GetProducts.html

CLI
- https://awscli.amazonaws.com/v2/documentation/api/latest/reference/pricing/get-products.html

```
❯ cat > filters.json
[
  {
    "Type": "TERM_MATCH",
    "Field": "ServiceCode",
    "Value": "AmazonEC2"
  },
  {
    "Type": "TERM_MATCH",
    "Field": "volumeType",
    "Value": "Provisioned IOPS"
  }
]
```

```
❯ aws pricing get-products --filters file://filters.json --format-version aws_v1 --max-results 1 --service-code AmazonEC2
```

[Output (AWS CLI output 형식을 YAML로 설정했는데, 내부의 `PriceList` 필드의 값은 JSON으로 옴)]
```yaml
FormatVersion: aws_v1
NextToken: uNv5M23uc+kwUioW6Wel4w==:CgQbPOZbIBCL/AEIKBWDjj7qFLXU3KRNrSbiJTrlDmkRWMu/uoI2rUfpysa//abdTIwTZ3nabZ/1l3QlTuNkEkfOxZt5fUaNZqGyl9L2RN2qsZRD94MCMNwlc86U3ogs5rJu0DiQkWjNYq
ir5qvWKfQkRUSqlokWU89vIdTGskkbgLvwTeoTolhavtbyWz3a
PriceList:
- '{"product":{"productFamily":"Storage","attributes":{"maxThroughputvolume":"1000
  MiB/s","volumeType":"Provisioned IOPS","maxIopsvolume":"64000","usagetype":"APS1-EBS:VolumeUsage.piops","locationType":"AWS
  Region","maxVolumeSize":"16 TiB","storageMedia":"SSD-backed","regionCode":"ap-southeast-1","servicecode":"AmazonEC2","volumeApiName":"io1","location":"Asia
  Pacific (Singapore)","servicename":"Amazon Elastic Compute Cloud","operation":""},"sku":"3MKHN58N7RDDVGKJ"},"serviceCode":"AmazonEC2","terms":{"OnDemand":{"3MKHN58N7RDDVGKJ.JRTCKXETXF":{"priceDimensions":{"3MKHN58N7RDDVGKJ.JRTCKXETXF.6YS6EN2CT7":{"unit":"GB-Mo","endRange":"Inf","description":"$0.138
  per GB-month of Provisioned IOPS SSD (io1)  provisioned storage - Asia Pacific (Singapore)","appliesTo":[],"rateCode":"3MKHN58N7RDDVGKJ.JRTCKXETXF.6YS6EN2CT7","beginRange":"0","pricePerUnit":{"USD":"0.1380000000"}}},"sku":"3MKHN58N7RDDVGKJ","effectiveDate":"2022-03-01T00:00:00Z","offerTermCode":"JRTCKXETXF","termAttributes":{}}}},"version":"20220322190035","publicationDate":"2022-03-22T19:00:35Z"}'
```

Filters 의 `Field` 와 `Value` 에 넣을 수 있는 값 목록을 알 수 있으면 좋겠네요.. 😊
